lib/token.js

```diff
@@ -151,87 +182,94 @@ function addFragmentListener(sdk, windowEl, timeout) {
     }
 
-    var clientId = oauthParams.clientId || sdk.options.clientId;
-    oauthUtil.validateClaims(sdk, jwt.payload, clientId, urls.issuer);
-
-    var idToken = {
-      idToken: res['id_token'],
-      claims: jwt.payload,
-      expiresAt: jwt.payload.exp,
-      scopes: scopes,
-      authorizeUrl: urls.authorizeUrl,
-      issuer: urls.issuer
-    };
+    if (res.state !== oauthParams.state) {
+      throw new AuthSdkError('OAuth flow response state doesn\'t match request state');
+    }
+    
+    if (res['access_token']) {
+      var accessToken = {
+        accessToken: res['access_token'],
+        expiresAt: Number(res['expires_in']) + Math.floor(Date.now()/1000),
+        tokenType: res['token_type'],
+        scopes: scopes,
+        authorizeUrl: urls.authorizeUrl,
+        userinfoUrl: urls.userinfoUrl
+      };
 
-    if (Array.isArray(tokenTypes)) {
-      tokenDict['id_token'] = idToken;
-    } else {
-      return idToken;
+      if (Array.isArray(tokenTypes)) {
+        tokenDict['token'] = accessToken;
+      } else {
+        return accessToken;
+      }
     }
-  }
-  
-  if (res['access_token']) {
-    var accessToken = {
-      accessToken: res['access_token'],
-      expiresAt: Number(res['expires_in']) + Math.floor(Date.now()/1000),
-      tokenType: res['token_type'],
-      scopes: scopes,
-      authorizeUrl: urls.authorizeUrl,
-      userinfoUrl: urls.userinfoUrl
-    };
 
-    if (Array.isArray(tokenTypes)) {
-      tokenDict['token'] = accessToken;
-    } else {
-      return accessToken;
+    if (res['code']) {
+      var authorizationCode = {
+        authorizationCode: res['code']
+      };
+
+      if (Array.isArray(tokenTypes)) {
+        tokenDict['code'] = authorizationCode;
+      } else {
+        return authorizationCode;
+      }
     }
-  }
 
-  if (res['code']) {
-    var authorizationCode = {
-      authorizationCode: res['code']
-    };
+    if (res['id_token']) {
+      var jwt = sdk.token.decode(res['id_token']);
+
+      var idToken = {
+        idToken: res['id_token'],
+        claims: jwt.payload,
+        expiresAt: jwt.payload.exp,
+        scopes: scopes,
+        authorizeUrl: urls.authorizeUrl,
+        issuer: urls.issuer,
+        clientId: clientId
+      };
 
-    if (Array.isArray(tokenTypes)) {
-      tokenDict['code'] = authorizationCode;
-    } else {
-      return authorizationCode;
+      return verifyToken(sdk, idToken, oauthParams.nonce, true)
+      .then(function(token) {
+        if (Array.isArray(tokenTypes)) {
+          tokenDict['id_token'] = idToken;
+        } else {
+          return idToken;
+        }
+      });
+    }
+  })
+  .then(function(codeOrToken) {
+    if (codeOrToken) {
+      return codeOrToken;
     }
-  }
 
-  if (!tokenDict['token'] && !tokenDict['id_token']) {
-    throw new AuthSdkError('Unable to parse OAuth flow response');
-  }
+    if (!tokenDict['token'] && !tokenDict['id_token']) {
```

Comment:
Not a big fan of tracking the tokenDict state outside of the promise chain approach.  Couple alternatives:

1. (preferred) Don't worry about response format in the preceding block. Always do the checks, and return an array of the tokens that you've extracted. In this block, just do an array sort on the tokens (by the tokenTypes), or return the token itself if the tokenType is not an array. 

2. Create a new tokenDict in the preceding promise block, and then return that, and accept it as an argument here.